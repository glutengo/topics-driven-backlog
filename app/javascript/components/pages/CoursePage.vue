<template>
    <section id="course">
        <template v-if="course">
            <h1>
                {{ course.title }}
                <small class="text-muted">{{ getSemesterInfo(course.semester_type, course.semester_year).fullString }}
                </small>
            </h1>

            <p>
                <a v-if="course.hyperlink" :href="course.hyperlink">{{ course.hyperlink }}</a>
            </p>

            <h2>Sprints</h2>

            <ul class="list-unstyled" v-if="sprints.length">
                <sprint class="mb-2" v-for="sprint in sprints" :data="sprint" :key="sprint.id"></sprint>
            </ul>

            <p v-else>No sprints in this course yet.</p>

            <sprint-form v-if="showForm" v-model="newSprint" @cancel="showForm = false" @submit="addSprint"></sprint-form>

            <b-button v-else type="button" variant="primary" @click="showForm = true">Add sprint</b-button>

        </template>
        <not-found v-else></not-found>
    </section>
</template>

<script>
    import NotFound from "../elements/NotFound.vue";
    import {info} from '../../helper/semester';
    import Sprint from "../elements/Sprint.vue";
    import SprintForm from "../forms/SprintForm.vue";

    export default {
        components: {SprintForm, Sprint, NotFound},
        data() {
            return {
                showForm: false,
                newSprint: {},
            };
        },
        computed: {
            course() {
                return this.$store.getters['courses/byId'](parseInt(this.$route.params.id, 10));
            },
            sprints() {
                return this.course.sprints ? this.course.sprints.sort((a, b) => a.start_date > b.start_date) : [];
            }
        },
        methods: {
            /**
             * Add a sprint to the given course
             */
            addSprint() {
                this.$store.dispatch('courses/addSprint', {
                    id: this.course.id,
                    sprint: this.newSprint
                }).then(() => {
                    this.newSprint = {};
                });
            },
            /**
             * Semester info for given type and year
             * @param type
             * @param year
             * @returns {{semesterStart: *, semester: string, year: number, fullString: string, valueString: string}}
             */
            getSemesterInfo(type, year) {
                return info(type, year);
            },
        },
    };
</script>

<style scoped>

</style>
