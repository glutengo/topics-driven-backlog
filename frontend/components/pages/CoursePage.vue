<template>
    <section id="course-page">
        <template v-if="course">
            <h1>
                {{ course.short_title }}
                <small class="text-muted">
                    {{ course.title }}
                    ({{ getSemesterInfo(course.semester_type, course.semester_year).fullString }})
                </small>
            </h1>

            <p v-if="course.hyperlink" >
                <a :href="course.hyperlink">{{ course.hyperlink }}</a>
            </p>

            <BCard no-body>
                <BTabs card>
                    <BTab title="Info">
                        Enrollment for this course is currently {{ course.allow_enrollment ? 'open' : 'closed' }}.
                    </BTab>
                    <BTab 
                        title="Projects" 
                        active>

                        <ul :class="['list-unstyled', $style.list]">
                            <ProjectItem
                                class="mb-2"
                                v-for="project in projects"
                                :data="project"
                                :key="project.id"/>

                            <ProjectItem
                                v-if="course.permissions.projects.create && !isEnrolledToProjectInCourse"
                                view="new"
                                :class="$style.new"
                                :data="newProject"
                                :course-id="course.id"/>
                        </ul>
                    </BTab>
                    <BTab 
                        v-if="course.permissions.topics.read" 
                        title="Topics">

                        <ul
                            class="list-unstyled"
                            v-if="topics.length">

                            <TopicItem
                                class="mb-2"
                                v-for="topic in topics"
                                :data="topic"
                                :key="topic.id"/>
                        </ul>

                        <p v-else>No topics in this course yet.</p>

                        <TopicForm
                            ref="topicForm"
                            v-if="showTopicForm"
                            v-model="newTopic"
                            @cancel="showTopicForm = false"
                            @submit="addTopic"/>

                        <BButton
                            v-else-if="course.permissions.topics.create"
                            type="button"
                            variant="primary"
                            @click="showTopicForm = true">Add topic</BButton>
                    </BTab>
                    <BTab
                        v-if="course.permissions.sprints.read" 
                        title="Sprints">

                        <ul 
                            class="list-unstyled" 
                            v-if="sprints.length">

                            <SprintItem 
                                class="mb-2" 
                                v-for="sprint in sprints" 
                                :data="sprint" 
                                :key="sprint.id"/>
                        </ul>

                        <p v-else>No sprints in this course yet.</p>

                        <SprintForm 
                            ref="sprintForm"
                            v-if="showSprintForm"
                            :course="course"
                            v-model="newSprint" 
                            @cancel="showSprintForm = false" 
                            @submit="addSprint"/>

                        <SprintCollectionForm 
                            ref="sprintCollectionForm"
                            v-else-if="showCollectionForm" 
                            v-model="newCollection"
                            @cancel="showCollectionForm = false"
                            @submit="addCollection"/>

                        <template v-else>
                            <BButton 
                                type="button" 
                                variant="primary" 
                                @click="showSprintForm = true">Add sprint</BButton>
                            
                            <BButton 
                                type="button" 
                                variant="primary" 
                                @click="showCollectionForm = true">Add multiple sprints</BButton>
                        </template>
                    </BTab>
                </BTabs>
            </BCard>
        </template>
        <NotFound v-else/>
    </section>
</template>

<script>
import BCard from '@bootstrap/card/card';
import BTabs from '@bootstrap/tabs/tabs';
import BTab from '@bootstrap/tabs/tab';
import BButton from '@bootstrap/button/button';
import NotFound from '@/components/elements/NotFound';
import SprintItem from '@/components/elements/SprintItem';
import SprintForm from '@/components/forms/SprintForm';
import SprintCollectionForm from '@/components/forms/SprintCollectionForm';
import ProjectForm from '@/components/forms/ProjectForm';
import ProjectItem from '@/components/elements/ProjectItem';
import TopicForm from '@/components/forms/TopicForm';
import TopicItem from '@/components/elements/TopicItem';
import { info } from '@/helper/semester';

export default {
    components: {
        SprintCollectionForm,
        SprintForm,
        ProjectForm,
        TopicForm,
        SprintItem,
        ProjectItem,
        TopicItem,
        NotFound,
        BCard,
        BTabs,
        BTab,
        BButton,
    },
    data() {
        return {
            showSprintForm: false,
            showCollectionForm: false,
            showProjectForm: false,
            showTopicForm: false,
            newSprint: this.$store.getters['sprints/template'](),
            newCollection: {},
            newProject: this.$store.getters['projects/template'](),
            newTopic: this.$store.getters['topics/template'](),
        };
    },
    computed: {
        /**
         * Course to display
         * @returns {object}
         */
        course() {
            return this.$store.getters['courses/byId'](
                parseInt(this.$route.params.id, 10)
            );
        },

        /**
         * Sprints in course sorted by start date
         * @returns {array}
         */
        sprints() {
            return this.course
                ? this.$store.getters['sprints/all'](this.course.id)
                    .sort((a, b) => a.start_date.localeCompare(b.start_date))
                : [];
        },

        /** 
         * Projects in course
         * @returns {array}
         */
        projects() {
            return this.course
                ? this.$store.getters['projects/all'](this.course.id)
                : [];
        },

        /**
         * Topics in course
         * @returns {array}
         */
        topics() {
            return this.course
                ? this.$store.getters['topics/all'](this.course.id)
                : [];
        },

        isEnrolledToProjectInCourse() {
            return this.projects.filter(
                p => p.user_ids && p.user_ids.indexOf(this.$store.state.user.id) > -1
            ).length > 0;
        },
    },
    methods: {
        /**
         * Add a sprint to the given course
         */
        async addSprint() {
            try {
                await this.$store.dispatch('sprints/create', {
                    parentId: this.course.id,
                    ...this.newSprint,
                });
    
                this.newSprint = this.$store.getters['sprints/template']();
                
                this.$nextTick(() => {
                    !this.$refs.sprintForm || this.$refs.sprintForm.$el.reset();
                });
            } catch (err) {
                /* istanbul ignore next */
                this.$notify({
                    title: 'Sprint creation failed',
                    text: err.body.message,
                    type: 'error',
                });
            }
        },

        /**
         * Add a sprint collection to the given course
         */
        async addCollection() {
            try {
                await this.$store.dispatch('sprints/createCollection', {
                    parentId: this.course.id,
                    collection: this.newCollection,
                });
    
                this.newCollection = {};

                this.$nextTick(() => {
                    !this.$refs.sprintCollectionForm || this.$refs.sprintCollectionForm.$el.reset();
                });
            } catch (err) {
                /* istanbul ignore next */
                this.$notify({
                    title: 'Sprints creation failed',
                    text: err.body.message,
                    type: 'error',
                });
            }
        },

        async addProject() {
            try {
                await this.$store.dispatch('projects/create', {
                    parentId: this.course.id,
                    ...this.newProject,
                });

                this.newProject = this.$store.getters['projects/template']();

                this.$nextTick(() => {
                    !this.$refs.projectForm || this.$refs.projectForm.$el.reset();
                });
            } catch (err) {
                /* istanbul ignore next */
                this.$notify({
                    title: 'Validation failed',
                    text: err.body.message.replace('Validation failed: ', ''),
                    type: 'error',
                });
            }
        },

        async addTopic() {
            try {
                await this.$store.dispatch('topics/create', {
                    parentId: this.course.id,
                    ...this.newTopic,
                });
    
                this.newTopic = this.$store.getters['topics/template']();      
                
                this.$nextTick(() => {
                    !this.$refs.topicForm || this.$refs.topicForm.$el.reset();
                });
            } catch (err) {                
                /* istanbul ignore next */
                this.$notify({
                    title: 'Topic creation failed',
                    text: err.body.message,
                    type: 'error',
                });
            }
        },

        /**
         * Semester info for given type and year
         * @param {string} type
         * @param {string} year
         * @returns {{semesterStart: *, semester: string, year: number, fullString: string, valueString: string}}
         */
        getSemesterInfo(type, year) {
            return info(type, year);
        },
    },
};
</script>

<style lang="scss" module>
    .list {
        display: grid;
        grid-gap: 8px;
        grid-template-columns: 1fr 1fr 1fr;
        grid-auto-rows: minmax(200px, 1fr);
    }

    .new {
        color: rgba(0, 0, 0, .125);
        border: 1px solid currentColor;

        &:hover {
            color: rgba(0, 0, 0, .25);
        }
    }
</style>