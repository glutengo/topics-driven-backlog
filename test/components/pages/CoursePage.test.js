import { mount } from 'vue-test-utils';
import mockRouter from 'mock-vue-router';
import CoursePage from '@/components/pages/CoursePage';

const { $router, $route } = mockRouter(['/courses/:id'], '/courses/1');

describe('CoursePage.test.js', () => {
    let cmp, dispatch = jest.fn();

    beforeEach(() => {
        cmp = mount(CoursePage, {
            mocks: {
                $router,
                $route,
                $store: {
                    dispatch,
                    state: {
                        loggedIn: true,
                        user: {
                            id: 1,
                        },
                    },
                    getters: {
                        'sprints/template': () => { return {}; },
                        'projects/template': () => { return {}; },
                        'topics/template': () => { return {}; },
                        'courses/byId': () => { 
                            return { 
                                id: 1,
                                title: 'Test course',
                                hyperlink: 'http://google.com',
                                short_title: 'TEST',
                                semester_type: 'WS',
                                semester_year: 2222,
                                permissions: {
                                    projects: {
                                        read: true,
                                        create: true,
                                    },
                                    topics: {
                                        read: true,
                                        create: true,
                                    },
                                    sprints: {
                                        read: true,
                                        create: true,
                                    },
                                },
                            }; 
                        },
                        'sprints/all': () => [
                            { id: 1, start_date: '2000-01-02', topic_ids: [] },
                            { id: 2, start_date: '2000-01-01', topic_ids: [] },
                        ],
                        'projects/all': () => [
                            { id: 1, title: 'Test', user_ids: [] },
                            { id: 2, title: 'Test 2', user_ids: [] },
                        ],
                        'topics/all': () => [],
                    },
                },
                $style: {},
            },
        });

        cmp.vm.$router.push('/courses/1');

        jest.resetAllMocks();
    });

    it('returns correct course from route param', () => {
        expect(cmp.vm.course.title).toBe('Test course');
    });

    it('returns correctly sorted sprints and projects for course', () => {
        expect(cmp.vm.sprints).toEqual([
            { id: 2, start_date: '2000-01-01', topic_ids: [] },
            { id: 1, start_date: '2000-01-02', topic_ids: [] },
        ]);

        expect(cmp.vm.projects).toEqual([
            { id: 1, title: 'Test', user_ids: [] },
            { id: 2, title: 'Test 2', user_ids: [] },
        ]);
    });

    it('adds a sprint and resets form', () => {
        cmp.vm.addSprint();

        expect(dispatch).toBeCalled();
        expect(cmp.vm.newSprint).toEqual({});
    });

    it('adds a sprint collection and resets form', () => {
        cmp.vm.addCollection();

        expect(dispatch).toBeCalled();
        expect(cmp.vm.newCollection).toEqual({});
    });


    it('adds a project and resets form', () => {
        cmp.vm.addProject();

        expect(dispatch).toBeCalled();
        expect(cmp.vm.newSprint).toEqual({});
    });

    it('adds a topic and resets form', () => {
        cmp.vm.addTopic();

        expect(dispatch).toBeCalled();
        expect(cmp.vm.newTopic).toEqual({});
    });

    it('does not calculate sprints, projects or topics if no course set', () => {
        cmp.vm.$store.getters['courses/byId'] = () => null;
        cmp.update();

        expect(cmp.vm.sprints.length).toBe(0);
        expect(cmp.vm.projects.length).toBe(0);
        expect(cmp.vm.topics.length).toBe(0);
    });
});
