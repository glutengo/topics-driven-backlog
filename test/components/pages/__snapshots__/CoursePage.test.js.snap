// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursePage.test.js has the expected html structure 1`] = `
<section
  id="course-page"
>
  <h1>
    
            Test course
            
    <small
      class="text-muted"
    >
      WS 17/18
            
    </small>
  </h1>
   
  <!---->
   
  <div
    class="card"
  >
    <div
      class="tabs"
      id="__BVID__59_"
    >
      <!---->
      <div
        class="card-header"
      >
        <ul
          class="nav nav-tabs card-header-tabs"
          id="__BVID__59___BV_tab_controls_"
          role="tablist"
          tabindex="0"
        >
          <li
            class="nav-item"
            role="presentation"
          >
            <a
              aria-controls="__BVID__59___BV_tab_container_"
              aria-posinset="1"
              aria-selected="true"
              aria-setsize="3"
              class="nav-link active"
              href="#"
              id="__BVID__60____BV_tab_button__"
              role="tab"
              tabindex="-1"
            >
              Projects
            </a>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <a
              aria-controls="__BVID__59___BV_tab_container_"
              aria-posinset="2"
              aria-selected="false"
              aria-setsize="3"
              class="nav-link"
              href="#"
              id="__BVID__62____BV_tab_button__"
              role="tab"
              tabindex="-1"
            >
              Topics
            </a>
          </li>
          <li
            class="nav-item"
            role="presentation"
          >
            <a
              aria-controls="__BVID__59___BV_tab_container_"
              aria-posinset="3"
              aria-selected="false"
              aria-setsize="3"
              class="nav-link"
              href="#"
              id="__BVID__64____BV_tab_button__"
              role="tab"
              tabindex="-1"
            >
              Sprints
            </a>
          </li>
        </ul>
      </div>
      <div
        class="tab-content"
        id="__BVID__59___BV_tab_container_"
      >
        <div
          aria-expanded="true"
          aria-hidden="false"
          aria-lablelledby="__BVID__60____BV_tab_button__"
          class="tab-pane card-body fade active v-enter v-enter-active"
          id="__BVID__60_"
          role="tabpanel"
          style=""
        >
          <p>
            No projects in this course yet.
          </p>
        </div>
         
        <div
          aria-expanded="false"
          aria-hidden="true"
          aria-lablelledby="__BVID__62____BV_tab_button__"
          class="tab-pane card-body fade"
          id="__BVID__62_"
          role="tabpanel"
          style="display: none;"
        >
          <p>
            No topics in this course yet.
          </p>
        </div>
         
        <div
          aria-expanded="false"
          aria-hidden="true"
          aria-lablelledby="__BVID__64____BV_tab_button__"
          class="tab-pane card-body fade"
          id="__BVID__64_"
          role="tabpanel"
          style="display: none;"
        >
          <p>
            No sprints in this course yet.
          </p>
           
          <button
            class="btn btn-primary"
            type="button"
          >
            Add sprint
          </button>
           
          <button
            class="btn btn-primary"
            type="button"
          >
            Add multiple sprints
          </button>
        </div>
        <!---->
      </div>
    </div>
  </div>
</section>
`;
